crucible/src/Lang/Crucible/CFG/Expr.hs

```diff
@@ -983,16 +977,16 @@ data App (ext :: Type) (f :: CrucibleType -> Type) (tp :: CrucibleType) where
   -- If no such position exists, return a negative value.
   StringIndexOf :: !(f (StringType si))
                 -> !(f (StringType si))
-                -> !(f NatType)
+                -> !(f IntegerType)
```

Comment:
In crucible/src/Lang/Crucible/Simulator/Evaluation.hs, should the `StringIsIndexOf` evaluation verify or add an assertion that the index argument is positive?  I don't know if the solver automatically does this check but for a concrete value it seems like we'd need something along the lines of:

```
k'' <- evalSub k
k' <- case asInt k'' of
        Nothing -> return k''  -- symbolic index
        Just n -> if n < 0
                  then addFailedAssertion sym $ GenericSimError "negative string index supplied to StringIndexOf"
                  else return k''
```

Reply:
This is a good point.  I'll check on what is happening here and decide what to do.

Reply:
Looking into it a bit more closely, the SMTLib functions being modeled here are actually total and have a well-defined result even for out-of-range and negative values of these indices. As such, I think I'm going to leave the crucible definitions here as-is and open a What4 ticket about ensuring these functions are faithful in the out-of-range cases.

Reply:
Might be worth adding a comment to that effect here.